uni/lib/controller/background_workers/notifications.dart

```diff
@@ -0,0 +1,150 @@
+import 'package:uni/controller/local_storage/app_shared_preferences.dart';
+import 'package:uni/controller/local_storage/notification_timeout_storage.dart';
+import 'package:uni/controller/networking/network_router.dart';
+import 'package:uni/model/entities/session.dart';
+import 'package:workmanager/workmanager.dart';
+
+///
+/// Stores all notifications that will be checked and displayed in the background. 
+/// Add your custom notification here, because it will NOT be added at runtime.
+/// (due to background worker limitations).
+/// 
+Map<Type, Notification Function()> notificationMap = {
+  TuitionNotification:() => TuitionNotification(),
+};
+
+
+
+
+abstract class Notification{
+  
+  String uniqueID;
+  Duration timeout;
+
+  Notification(this.uniqueID, this.timeout);
+
+  Future<Tuple2<String, String>> buildNotificationContent(Session session);
+
+  Future<bool> checkConditionToDisplay(Session session);
+
+  void displayNotification(Tuple2<String, String> content, FlutterLocalNotificationsPlugin localNotificationsPlugin);
+
+  Future<void> displayNotificationIfPossible(Session session, FlutterLocalNotificationsPlugin localNotificationsPlugin) async{
+    final bool test = await checkConditionToDisplay(session);
+    if(test){
+      displayNotification(await buildNotificationContent(session), localNotificationsPlugin);
+    }
+  }
+}
+
+class NotificationManager{
+
+
+  static final FlutterLocalNotificationsPlugin _localNotificationsPlugin = FlutterLocalNotificationsPlugin();
+
+  static bool _initialized = false;
+
+  static const Duration _notificationWorkerPeriod = Duration(hours: 1);
+
+
+  static Future<void> updateAndTriggerNotifications() async{
+    //first we get the .json file that contains the last time that the notification have ran
+    _initFlutterNotificationsPlugin();
+    final notificationStorage = await NotificationTimeoutStorage.create();
+    final userInfo = await AppSharedPreferences.getPersistentUserInfo();
+    final faculties = await AppSharedPreferences.getUserFaculties();
+
+    final Session session =  await NetworkRouter.login(userInfo.item1, userInfo.item2, faculties, false);
+
+    for(Notification Function() value in notificationMap.values){
+        final Notification notification = value();
+        final DateTime lastRan = notificationStorage.getLastTimeNotificationExecuted(notification.uniqueID);
+        if(lastRan.add(notification.timeout).isBefore(DateTime.now())) {
+          await notification.displayNotificationIfPossible(session, _localNotificationsPlugin);
+          notificationStorage.addLastTimeNotificationExecuted(notification.uniqueID, DateTime.now());
+      }
+    }
+  }
+
+  static void initializeNotifications() async{
+    //guarentees that the execution is only done once in the lifetime of the app. 
+    if(_initialized) return;
+    _initFlutterNotificationsPlugin();
+    _buildNotificationWorker();
+    _initialized = true;
```

Comment:
Perhaps this assignment should come before the other two tasks to avoid some race conditions

Reply:
I agree, done